[Documentation] Quick reference guide for rocprofv3 (#953)
ROCm/rocm-systems: projects/rocprofiler-sdk/source/docs/how-to/using-rocpd-output-format.rst

   /opt/rocm/bin/rocpd convert -i db{3,4}.db --output-format csv otf2 pftrace

Dedicated Conversion Tools

> ++++++++++++++++++++++++++

Reviewer: In addition to adding this section, I was hoping we add a "Summary" section, and perhaps a section to highlight "Aggregating rocpd data".  You may also want to add a "Query" section since that is another module we added now too.  Like, when they do `rocpd --help`, they'll see {convert, query, summary}.

Summary: basically is providing the same summaries that rocprofv3 used to provide, but using the DB data.  So you can link to https://rocm.docs.amd.com/projects/rocprofiler-sdk/en/latest/how-to/using-rocprofv3.html#summary, to explain.

Aggregation: We should call out that rocpd allows multiple databases (runs on different ranks/nodes) can be combined together for analysis, something we couldn't do with rocprofv3 perfetto output before.  

===
Also, rocpd Summary has an added benefit, in that, it can aggregate multiple DB data, and we want to highlight that.  If we have multiple databases, you can summarize them together, using the built in summary reports.  BUT if you want to analyze them by rank, you can use `--summary-by-rank` to get an analysis comparing one rank/node to another.

Author: Done

ROCprofiler-SDK provides specialized conversion utilities for efficient format-specific operations. These tools offer streamlined interfaces for single-format conversions and are particularly useful in automated workflows and scripts.

> rocpd2csv - CSV Export Tool

Reviewer: I didn't think we needed to go into so much detail for each of these "direct conversion" scripts, but I guess we can keep them for now.

Author: Sure, I think we should keep them as well for now

**Key Features:**

- **Multi-Format Output:** Supports console, CSV, HTML, JSON, Markdown, and PDF report generation
- **Comprehensive Statistics:** Kernel execution times, API call frequencies, memory transfer analysis
- **Domain-Specific Analysis:** Separate summaries for HIP, ROCr, Markers, and other trace domains
- **Rank-Based Analysis:** Per-process and per-rank performance breakdowns for MPI applications
- **Configurable Scope:** Selective inclusion/exclusion of analysis categories

**Output Format Options:**

.. code-block:: bash

   # Console output (default)
   rocpd2summary -i profile.db

   # CSV format for data analysis
   rocpd2summary -i profile.db --format csv -o performance_metrics

   # HTML report with visualization
   rocpd2summary -i profile.db --format html -d ~/reports/

   # Multiple output formats
   rocpd2summary -i profile.db --format csv html json

**Analysis Categories:**

.. code-block:: bash

   # Include all available domains
   rocpd2summary -i profile.db --region-categories HIP ROCR MARKERS KERNEL

   # Focus on GPU kernel analysis only
   rocpd2summary -i profile.db --region-categories KERNEL

   # Exclude markers to speed up processing

>    rocpd2summary -i profile.db --region-categories HIP ROCR KERNEL

Reviewer: ```suggestion
   rocpd2summary -i profile.db --region-categories HIP HSA MARKERS KERNEL

   # Focus on GPU kernel analysis only
   rocpd2summary -i profile.db --region-categories KERNEL

   # Exclude markers to speed up processing
   rocpd2summary -i profile.db --region-categories HIP HSA KERNEL
```
Do we actually have a ROCR API category?  When I do sys-trace, I usually just get: HIP HSA MARKER categories only.

Author: No! you are right. I have removed the `ROCr` mention.

**Advanced Analysis Options:**

.. code-block:: bash

   # Include domain-specific statistics
   rocpd2summary -i multi_gpu.db --domain-summary

   # Per-rank analysis for MPI applications
   rocpd2summary -i mpi_profile_*.db --summary-by-rank --format html

   # Time-windowed summary analysis
   rocpd2summary -i long_run.db --start 25% --end 75% --format csv

**Report Content:**
- **Kernel Statistics:** Execution time distributions, call frequencies, grid/block sizes
- **API Timing:** HIP/ROCr function call durations and frequencies
- **Memory Analysis:** Transfer patterns, bandwidth utilization, allocation statistics  
- **Device Utilization:** GPU occupancy patterns and idle time analysis
- **Synchronization Overhead:** Barrier and synchronization point analysis

**Output Files:**
- ``summary_kernel.{format}`` - GPU kernel execution summary
- ``summary_hip_api.{format}`` - HIP API call statistics

> - ``summary_rocr_api.{format}`` - ROCr runtime API analysis

Reviewer: I see you have ROCr here too, maybe the workload you ran has ROCr api traces.  If so, you can ignore my above comment.

Author: No! you are right. I have removed the ROCr mention.

   
   # Cross-rank performance comparison for MPI applications
   rocpd summary -i rank_0.db rank_1.db rank_2.db rank_3.db --summary-by-rank
   
   # Multi-node scaling analysis
   rocpd query -i node_*.db \
               --query "SELECT COUNT(*) as total_kernels, SUM(duration) as total_time FROM kernels"

**Distributed Computing Workflows**

**MPI Application Analysis:**

.. code-block:: bash

   # Profile MPI application across multiple ranks
   mpirun -np 4 rocprofv3 --hip-trace --output-format rocpd -- mpi_application
   
   # Generate aggregated performance summary
   rocpd summary -i results_rank_*.db --summary-by-rank --format html -o mpi_performance_report
   
   # Analyze load balancing across ranks
   rocpd query -i results_rank_*.db \
               --query "SELECT pid, COUNT(*) as kernel_count, AVG(duration) as avg_duration FROM kernels GROUP BY pid"

**Multi-GPU Scaling Analysis:**

.. code-block:: bash

   # Profile application with multiple GPU devices
   rocprofv3 --hip-trace --device 0,1,2,3 --output-format rocpd -- multi_gpu_app
   
   # Aggregate device utilization analysis
   rocpd query -i multi_gpu_results.db \

>                --query "SELECT device_id, COUNT(*) as operations, SUM(duration) as total_time FROM kernels GROUP BY device_id"

Reviewer: ```suggestion
               --query "SELECT agent_abs_index as device_id, COUNT(*) as operations, SUM(duration) as total_time FROM kernels GROUP BY device_id"
```
Found another one, can make it readable like this.

Author: Yes, makes sense

rocpd query - SQL Query Engine
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

**Purpose:** Execute custom SQL queries against rocpd databases with support for multiple output formats, automated reporting, and email delivery.

**Location:** ``/opt/rocm/bin/rocpd query``

**Syntax:**

.. code-block:: bash

   rocpd query -i INPUT [INPUT ...] --query "SQL_STATEMENT" [OPTIONS]

**Key Features:**

- **Standard SQL Support:** Full SQLite3 SQL syntax including JOINs, aggregate functions, and complex WHERE clauses
- **Multi-Database Aggregation:** Query across multiple database files as unified virtual database
- **Multiple Output Formats:** Console, CSV, HTML, JSON, Markdown, PDF, and interactive dashboards
- **Script Execution:** Execute complex SQL scripts with view definitions and custom functions
- **Automated Reporting:** Email delivery with SMTP configuration and attachment management
- **Time Window Integration:** Apply temporal filtering before query execution

> Database Schema and Views
> ~~~~~~~~~~~~~~~~~~~~~~~~~
> 
> rocpd databases provide comprehensive views for analysis:
> 

Reviewer: ```suggestion
Database Schema and Views
~~~~~~~~~~~~~~~~~~~~~~~~~

rocpd databases provide comprehensive views for analysis.  In general, any queries should be built using the `data_views`:

```
We want people to just use the data_views, not any of the other tables.  Maybe we point them to where the schema files are located in our repository so anyone interested can view the .sql files, to help them build queries?

https://github.com/ROCm/rocm-systems/blob/develop/projects/rocprofiler-sdk/source/share/rocprofiler-sdk-rocpd/data_views.sql

Author: done

**Core Data Views:**

.. code-block:: sql

   -- System and hardware information
   SELECT * FROM rocpd_info_agents;
   SELECT * FROM rocpd_info_node;
   
   -- Kernel execution data
   SELECT * FROM kernels;
   SELECT * FROM top_kernels;
   
   -- API trace information
   SELECT * FROM hip_api_trace;

>    SELECT * FROM rocr_api_trace;

Reviewer: ```suggestion
   SELECT * FROM regions_and_samples WHERE category LIKE 'HIP_%'
   SELECT * FROM regions_and_samples WHERE category LIKE 'RCCL_%'
```
We don't actually have the _API_TRACE views anymore.  They need a slightly more complicated query to get that data, like that ^

Author: done

   
   -- Performance counters
   SELECT * FROM counters_collection;
   
   -- Memory operations
   SELECT * FROM memory_copies;
   SELECT * FROM memory_allocations;
   
   -- Process and thread information
   SELECT * FROM processes;
   SELECT * FROM threads;
   
   -- Marker and region data
   SELECT * FROM regions;

>    SELECT * FROM markers;

Reviewer: ```suggestion
   SELECT * FROM regions_and_samples WHERE category LIKE 'MARKERS_%'
```
Markers comes from a query like this

Author: done

**Summary and Analysis Views:**

.. code-block:: sql

   -- Top performing kernels by execution time
   SELECT * FROM top_kernels LIMIT 10;
   
   -- Memory bandwidth analysis
   SELECT * FROM memory_copy_summary;
   
   -- API call frequency statistics
   SELECT * FROM api_call_statistics;
   
   -- Device utilization metrics

>    SELECT * FROM device_utilization;

Reviewer: These views/tables don't exist so those won't work.  We should remove these
```suggestion
 
```

>    -- Memory bandwidth analysis
>    SELECT * FROM memory_copy_summary;

Reviewer: ```suggestion
   -- Top analysis
   SELECT * FROM top;
   
   -- Busy analysis
   SELECT * FROM busy;
```
We also don't have memory_copy_summary.  All summary queries are dynamically generated in python (as requested by JM).  He doesn't want any summary views stored in the schema.  So all we have stored, are top, top_kernels, busy.

Author: Got it

**Simple Data Exploration:**

.. code-block:: bash

   # List available GPU agents
   rocpd query -i profile.db --query "SELECT * FROM rocpd_info_agents"
   
   # Show top 10 longest-running kernels
   rocpd query -i profile.db --query "SELECT name, duration FROM kernels ORDER BY duration DESC LIMIT 10"
   
   # Count total number of kernel dispatches
   rocpd query -i profile.db --query "SELECT COUNT(*) as total_kernels FROM kernels"

**Multi-Database Aggregation:**

.. code-block:: bash

   # Combine data from multiple profiling sessions
   rocpd query -i session1.db session2.db session3.db \
               --query "SELECT pid, COUNT(*) as kernel_count FROM kernels GROUP BY pid"
   
   # Cross-session performance comparison
   rocpd query -i baseline.db optimized.db \
               --query "SELECT name, AVG(duration) as avg_duration FROM kernels GROUP BY name"

**Advanced Analytics:**

.. code-block:: bash

   # Kernel performance analysis with statistics

>    rocpd query -i profile.db --query "
>    SELECT 
>        name,
>        COUNT(*) as dispatch_count,
>        MIN(duration) as min_duration,
>        AVG(duration) as avg_duration,
>        MAX(duration) as max_duration,
>        SUM(duration) as total_duration
>    FROM kernels 
>    GROUP BY name 
>    ORDER BY total_duration DESC"

Reviewer: ```suggestion
   rocpd query -i profile.db --query "
   SELECT 
       name as kernel_name,
       COUNT(*) as dispatch_count,
       MIN(duration) as min_duration,
       AVG(duration) as avg_duration,
       MAX(duration) as max_duration,
       SUM(duration) as total_duration
   FROM kernels 
   GROUP BY kernel_name
   ORDER BY total_duration DESC"
```
Just realized, we can do this, to make it more readable... like you had it before.  If you like it this way, then, feel free to make the changes to the rest of the queries.

Otherwise, just ignore this comment and just leave them.  Up to you.

Author: yes, its more readable

>    SELECT 
>        kernel_name,
>        COUNT(*) as dispatch_count,
>        MIN(duration_ns) as min_duration,
>        AVG(duration_ns) as avg_duration,
>        MAX(duration_ns) as max_duration,
>        SUM(duration_ns) as total_duration
>    FROM kernels 
>    GROUP BY kernel_name 
>    ORDER BY total_duration DESC"

Reviewer: ```suggestion
   SELECT 
       name,
       COUNT(*) as dispatch_count,
       MIN(duration) as min_duration,
       AVG(duration) as avg_duration,
       MAX(duration) as max_duration,
       SUM(duration) as total_duration
   FROM kernels 
   GROUP BY name 
   ORDER BY total_duration DESC"
```
Fixed the fields to match the kernel view column names.  This works, I ran the query.

Author: Thanks! Done

**Memory Transfer Analysis:**

.. code-block:: bash

>    # Memory copy performance by direction
>    rocpd query -i profile.db --query "
>    SELECT 
>        copy_kind,
>        COUNT(*) as transfer_count,
>        SUM(size_bytes) as total_bytes,
>        AVG(bandwidth_gb_per_sec) as avg_bandwidth
>    FROM memory_copies 
>    GROUP BY copy_kind 
>    ORDER BY total_bytes DESC"

Reviewer: ```suggestion
   # Memory copy analysis by direction
   rocpd query -i profile.db --query "
   SELECT 
       name,
       src_agent_type,
       src_agent_abs_index,
       dst_agent_type,
       dst_agent_abs_index,
       COUNT(*) as transfer_count,
       SUM(size) as total_bytes,
       SUM(duration) as total_duration
   FROM memory_copies 
   GROUP BY src_agent_abs_index 
   ORDER BY total_bytes DESC"
```
Changed this query to make it work.  Had to drop the avg_bandwidth because that's not stored by default.

   rocpd query -i profile.db --query "SELECT * FROM top_kernels" --format html -o performance_report

**Interactive Dashboard:**

.. code-block:: bash

   # Create interactive HTML dashboard
   rocpd query -i profile.db --query "SELECT * FROM device_utilization" --format dashboard -o utilization_dashboard
   
   # Use custom dashboard template
   rocpd query -i profile.db --query "SELECT * FROM kernels" --format dashboard \
               --template-path ~/templates/custom_dashboard.html -o custom_report

**JSON for Programmatic Integration:**

.. code-block:: bash

   # Export structured JSON data
   rocpd query -i profile.db --query "SELECT * FROM counters_collection" --format json -o counter_data

**PDF Reports:**

.. code-block:: bash

   # Generate PDF report with monospace formatting
   rocpd query -i profile.db --query "SELECT kernel_name, duration_ns FROM top_kernels" --format pdf -o kernel_report

Script-Based Analysis
~~~~~~~~~~~~~~~~~~~~~

Execute complex SQL scripts with view definitions and custom analysis logic:

**SQL Script Example (analysis.sql):**

.. code-block:: sql

   -- Create temporary views for complex analysis
   CREATE TEMP VIEW kernel_stats AS

>    SELECT 
>        kernel_name,
>        COUNT(*) as dispatch_count,
>        AVG(duration_ns) as avg_duration,
>        STDDEV(duration_ns) as duration_stddev
>    FROM kernels 
>    GROUP BY kernel_name;
>    
>    CREATE TEMP VIEW performance_outliers AS
>    SELECT k.*, ks.avg_duration, ks.duration_stddev
>    FROM kernels k
>    JOIN kernel_stats ks ON k.kernel_name = ks.kernel_name
>    WHERE ABS(k.duration_ns - ks.avg_duration) > 2 * ks.duration_stddev;

Reviewer: ```suggestion
   SELECT 
       name,
       COUNT(*) as dispatch_count,
       AVG(duration) as avg_duration,
       STDDEV(duration) as duration_stddev
   FROM kernels 
   GROUP BY name;
   
   CREATE TEMP VIEW performance_outliers AS
   SELECT k.*, ks.avg_duration, ks.duration_stddev
   FROM kernels k
   JOIN kernel_stats ks ON k.name = ks.name
   WHERE ABS(k.duration - ks.avg_duration) > 2 * ks.duration_stddev;
```

Fixed the query to use proper column names
